design/user-event-spi.md

```diff
@@ -0,0 +1,149 @@
+
+* **Status**: Draft v0.1
+* **Author**: [vmuzikar](https://github.com/vmuzikar)
+
+## Motivation
+
+Current Event SPI design has several limitations. E.g. it's not defined when an event is fired exactly, there are no pre-
+and post-events. There's also no way of pre-events intercepting for approve/reject functionality.
+
+## Use cases
+
+* A user is removed from Keycloak. All related records in an external system are removed before the user is permanently
+  deleted (e.g. eshop orders).
+* A user self-registers in Keycloak. Their registration must be approved by an external system (either human or automated)
+  before the new account is actually fully created.
+* A user self-edits their email address. The change must be approved first by an external system before it's actually
+  applied.
+
+## Design proposal
+
+To overcome the current Event SPI limitations, a new "domain-specific SPI" will be implemented – User Event SPI.
+
+Key changes over current generic Event SPI:
+* Each event is fired twice – as a pre- and post-event.
+* Pre-events are "interactive" – listener can approve, reject or delegate (for later decision by an external system)
+  each pre-event. 
+* The same User Event SPI will be used everywhere – admin console and REST endpoints, user registration form,
+  account console and REST endpoints, etc.
+* The event types will be much more detailed, not just usual CRUD. There'll be separated events for user self-registration,
+  creation through Admin REST API, Account Console, etc.
+
+### Terminology
+
+User Action (in this context) is a process which fires the User Event. This can be e.g. a user creation through Admin REST
+API, user self-registration, user field modification, assigning a user role, etc.
+
+Pre-event is an event that is fired after all other verifications passed (authorization, request validation, etc.), just
+before persisting any changes.
+
+Post-event is an event that is fired after a User Action has finished completely and all changes are persisted.
+
+### Event types
+
+* User is created manually through admin endpoints
+* Basic user info (name, email, etc.) is modified through admin endpoints
+* Basic user info (name, email, etc.) is modified through account endpoints
+* User self-registers through login page
+* User self-registers through an IdP, i.e. logs in for the first time
+* Admin adds a federation link
+* Admin removes a federation link
+* User logs in using federation for the first time
+* User is added to a group (through admin endpoints)
+* User is removed from a group (through admin endpoints)
+* User has a realm or client role assigned (through admin endpoints)
+* User has a realm or client role unassigned (through admin endpoints)
+* Admin updates a user's password through admin endpoints (not interactive)
+* User updates their password (not interactive)
+* User adds an authenticator
+* User removes an authenticator
+* Admin removes an authenticator from a user
+* User is deleted
+
```

Comment:
I miss few event types there we already use in our keycloak extension:
* auth flow initiates for user (keycloak obtained auth request by protocols like OIDC or SAML2 and starts auth flow)
* auth flow is finished for user (just before user's browser is redirected back to client app with OIDC or SAML2 etc)
* user performed successful authentication (by username/password form submit, social login or any other authenticator like SMS) - may be fired multiple times during auth flow if more factor auth is used
* required action performed/finished by user (including email verification, save of updated user info etc) - may be fired multiple times during auth flow if more required actions is execured

Reply:
The idea is to split the current event listener into multiple more bespoke apis. This one is for user crud operations. We'd add a different one for login/actions/etc.

Reply:
I see, that's true that we do not need pre and post for events I proposed, thanks